File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shaders/uniforms2/Vec3Uniform.java
package org.saar.lwjgl.opengl.shaders.uniforms2;

import org.joml.Vector3fc;
import org.lwjgl.opengl.GL20;

public abstract class Vec3Uniform extends UniformBase implements Uniform {

    @Override
    public final void doLoad() {
        final Vector3fc v = getUniformValue();
        GL20.glUniform3f(getLocation(), v.x(), v.y(), v.z());
    }

    public abstract Vector3fc getUniformValue();
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shader/uniforms/Vec3Uniform.kt
package org.saar.lwjgl.opengl.shader.uniforms

import org.joml.Vector3fc
import org.lwjgl.opengl.GL20

abstract class Vec3Uniform : Uniform {

    final override fun load(location: Int) {
        val value: Vector3fc = this.value
        GL20.glUniform3f(location, value.x(), value.y(), value.z())
    }

    abstract val value: Vector3fc
}